tempesta_fw/sock_srv.c

```diff
@@ -1248,262 +1699,528 @@ static int
+		if ((r = tfw_cfgop_start_sg_cfg(sg_cfg)))
+			return r;
+
+	tfw_sg_apply_reconfig(&orphan_sgs);
+	list_for_each_entry_safe(sg, tmp_sg, &orphan_sgs, list)
+		if ((r = tfw_sock_srv_grace_shutdown_sg(sg)))
+			return r;
+
+	tfw_http_sess_use_sticky_sess(tfw_cfg_use_sticky_sess);
+	tfw_cfgop_cleanup_srv_cfgs(false);
+
+	return 0;
 }
 
 static void
 tfw_sock_srv_stop(void)
 {
+	/* tfw_sock_srv_start() may failed just in the middle. */
+	tfw_sg_for_each_srv_reconfig(tfw_sock_srv_disconnect_srv);
+	tfw_sock_srv_grace_shutdown_now();
 	tfw_sg_for_each_srv(tfw_sock_srv_disconnect_srv);
+	tfw_sg_release_all();
 }
 
 static TfwCfgSpec tfw_srv_group_specs[] = {
 	{
-		"server", NULL,
-		tfw_cfgop_in_server,
+		.name = "server",
+		.deflt = NULL,
+		.handler = tfw_cfgop_in_server,
+		.cleanup = tfw_cfgop_cleanup_srv_groups,
 		.allow_repeat = true,
-		.cleanup = tfw_clean_srv_groups
+		.allow_reconfig = true,
 	},
 	{
-		"sched", "ratio static",
-		tfw_cfgop_in_sched,
+		.name = "sched",
+		.deflt = "ratio static",
+		.handler = tfw_cfgop_in_sched,
+		.cleanup = tfw_cfgop_cleanup_srv_groups,
 		.allow_none = true,
 		.allow_repeat = false,
-		.cleanup = tfw_clean_srv_groups,
+		.allow_reconfig = true,
 	},
 	{
-		"server_queue_size", "1000",
-		tfw_cfgop_in_queue_size,
+		.name = "server_queue_size",
+		.deflt = "1000",
+		.handler = tfw_cfgop_in_queue_size,
+		.cleanup = tfw_cfgop_cleanup_srv_groups,
+		.spec_ext = &(TfwCfgSpecInt) {
+			.range = { 0, INT_MAX },
+		},
+		.allow_none = true,
+		.allow_repeat = false,
+		.allow_reconfig = true,
+	},
+	{
+		.name = "server_forward_timeout",
+		.deflt = "60",
+		.handler = tfw_cfgop_in_fwd_timeout,
+		.cleanup = tfw_cfgop_cleanup_srv_groups,
+		.spec_ext = &(TfwCfgSpecInt) {
+			.range = { 0, INT_MAX },
+		},
 		.allow_none = true,
 		.allow_repeat = false,
-		.cleanup = tfw_clean_srv_groups,
+		.allow_reconfig = true,
+	},
+	{
+		.name = "server_forward_retries",
+		.deflt = "5",
+		.handler = tfw_cfgop_in_fwd_retries,
+		.cleanup = tfw_cfgop_cleanup_srv_groups,
 		.spec_ext = &(TfwCfgSpecInt) {
 			.range = { 0, INT_MAX },
 		},
+		.allow_none = true,
+		.allow_repeat = false,
+		.allow_reconfig = true,
 	},
 	{
-		"server_forward_timeout", "60",
-		tfw_cfgop_in_fwd_timeout,
+		.name = "server_retry_nonidempotent",
+		.deflt = TFW_CFG_DFLT_VAL,
+		.handler = tfw_cfgop_in_retry_nip,
+		.cleanup = tfw_cfgop_cleanup_srv_groups,
 		.allow_none = true,
 		.allow_repeat = false,
-		.cleanup = tfw_clean_srv_groups,
+		.allow_reconfig = true,
+	},
+	{
+		.name = "server_connect_retries",
+		.deflt = "10",
+		.handler = tfw_cfgop_in_conn_retries,
+		.cleanup = tfw_cfgop_cleanup_srv_groups,
 		.spec_ext = &(TfwCfgSpecInt) {
 			.range = { 0, INT_MAX },
 		},
+		.allow_none = true,
+		.allow_repeat = false,
+		.allow_reconfig = true,
+	},
+	{
+		.name = "sticky_sessions",
+		.deflt = TFW_CFG_DFLT_VAL,
+		.handler = tfw_cfgop_in_sticky_sess,
+		.cleanup = tfw_cfgop_cleanup_srv_groups,
```

Comment:
As I understand `cleanup` callbacks is not called for nested `srv_group` options specs. Does it make sense to define these callbacks?